node/src/template.rs

```diff
@@ -0,0 +1,52 @@
+use bitcoincore_rpc::{Auth, Client, RpcApi};
+use bitcoincore_rpc_json::{GetBlockTemplateModes, GetBlockTemplateResult, GetBlockTemplateRules};
+use std::error::Error;
+use tokio::sync::mpsc::{Receiver, Sender};
+use tokio::time::{sleep, Duration};
+use chrono::prelude::*;
+
+const BLOCK_TEMPLATE_RULES: [GetBlockTemplateRules; 4] = [
+    GetBlockTemplateRules::SegWit,
+    GetBlockTemplateRules::Signet,
+    GetBlockTemplateRules::Csv,
+    GetBlockTemplateRules::Taproot,
+];
+
+pub async fn template_poll(
+    rpc_url: String,
+    rpc_user: String,
+    rpc_pass: String,
+    poll_interval: u64,
+    template_tx: Sender<GetBlockTemplateResult>,
+) -> Result<(), Box<dyn Error + Send>> {
+    let rpc = Client::new(&rpc_url, Auth::UserPass(rpc_user, rpc_pass)).unwrap();
+    
+    let mut last_template = rpc
+        .get_block_template(GetBlockTemplateModes::Template, &BLOCK_TEMPLATE_RULES, &[])
+        .unwrap();
+    template_tx.send(last_template.clone()).await.unwrap();
+
+    loop {
+        // sleep until it's time to poll again
+        sleep(Duration::from_secs(poll_interval.clone())).await;
+
+        let template = rpc
+            .get_block_template(GetBlockTemplateModes::Template, &BLOCK_TEMPLATE_RULES, &[])
+            .unwrap();
```

Comment:
unwrap here and on L39 will kill the loop and the program. We should explicitly handle the error and then continue with the loop. The BIP also says we should do exponential backoff before trying again.

1. Handle error
2. Do exponential backoff with parameters as module constants
3. If after cumulative retry period we are still seeing an error then we should print meaningful error and do a proper shutdown. We don't have proper shutdown yet, so for now a process::exit will do. Added an issue for this https://github.com/braidpool/braidpool/issues/55

Reply:
I understand your point on proper error handling and I agree on the need for a more elegant strategy. I'm working on something.

But just to clarify, L39 would throw an error under different circumstances.
- `template = rpc.get_block_template(...).unwrap()` failure depends on the RPC result.
- `template_tx.send(template.clone()).await.unwrap()` failure depends on the consumer of the `mpsc` channel being dropped. For now I only wrote a dummy `fn consumer` which makes it unclear under which circumstances this scenario will arise.

Reply:
just proposed a solution for exponential backoff on https://github.com/braidpool/braidpool/pull/54/commits/98e88116717a5072fb8e567f6942883eb1f78876

the `unwrap` on the `mpsc` channel remains, however.

waiting for your input on this one @pool2win